# ANOVA

# Analysis of Variance (ANOVA)

ANOVA (Analysis of Variance) is a statistical method used for comparing the means of multiple populations. Previously, we have considered only a single population or at most two populations. A one-way ANOVA uses one independent variable, while a two-way ANOVA uses two independent variables. The statistical distribution used in ANOVA is the F-distribution, whose characteristics are as follows:

1. The F-distribution has a single tail (toward the right) and contains only positive values.

   <center><img src="data/f_dist.png"/></center>

2. The F-statistic, which is the critical statistic in ANOVA, is the ratio of variation between the sample means to the variation within the samples. The formula is as follows:
   $$F = \frac{\text{variation between sample means}}{\text{variation within the samples}}$$

3. The different populations are referred to as treatments.
4. A high value of the F-statistic implies that the variation between samples is considerable compared to the variation within the samples. In other words, the populations or treatments from which the samples are drawn are actually different from one another.
5. Random variations between treatments are more likely to occur when the variation within the sample is considerable.

Use a one-way ANOVA when you have collected data about one categorical independent variable and one quantitative dependent variable. The independent variable should have at least three levels (i.e., at least three different groups or categories).

ANOVA tells you if the dependent variable changes according to the level of the independent variable. For example:

+ Your independent variable is social media use, and you assign groups to low, medium, and high levels of social media use to find out if there is a difference in hours of sleep per night.
+ Your independent variable is the brand of soda, and you collect data on Coke, Pepsi, Sprite, and Fanta to find out if there is a difference in the price per 100ml.

ANOVA determines whether the groups created by the levels of the independent variable are statistically different by calculating whether the means of the treatment levels are different from the overall mean of the dependent variable. If any of the group means is significantly different from the overall mean, then the null hypothesis is rejected.

ANOVA uses the F-test for statistical significance. This allows for the comparison of multiple means at once, as the error is calculated for the whole set of comparisons rather than for each individual two-way comparison (which would happen with a t-test).

The F-test compares the variance in each group mean from the overall group variance. If the variance within groups is smaller than the variance between groups, the F-test will find a higher F-value, and therefore a higher likelihood that the difference observed is real and not due to chance.

The assumptions of the ANOVA test are the same as the general assumptions for any parametric test:

+ **Independence of observations:** The data were collected using statistically valid methods, and there are no hidden relationships among observations. If your data fail to meet this assumption because you have a confounding variable that you need to control for statistically, use an ANOVA with blocking variables.
+ **Normally distributed response variable:** The values of the dependent variable follow a normal distribution.
+ **Homogeneity of variance:** The variation within each group being compared is similar for every group. If the variances are different among the groups, then ANOVA probably isn’t the right fit for the data.


## One-Way-ANOVA

A few agricultural research scientists have planted a new variety of cotton called “AB
cotton.” They have used three different fertilizers – A, B, and C – for three separate
plots of this variety. The researchers want to find out if the yield varies with the type of
fertilizer used. Yields in bushels per acre are mentioned in the below table. Conduct an
ANOVA test at a 5% level of significance to see if the researchers can conclude that there
is a difference in yields.

| Fertilizer A | Fertilizer b | Fertilizer c |
|--------------|--------------|--------------|
|     40       |     45       |     55       |
|     30       |     35       |     40       |
|     35       |     55       |     30       |
|     45       |     25       |     20       |

Null hypothesis: $H_0 : \mu_1 = \mu_2 = \mu_3$  
Alternative hypothesis: $H_1$: At least one $\mu_i$ differs from the others (i.e., not all population means are equal)

the level of significance: $\alpha$=0.05

In [1]:
# ANOVA Assumptions (should be checked before running):
# 1. Normality: each group should be approximately normally distributed
# 2. Homogeneity of variance: group variances should be roughly equal (check with Levene's test)
# 3. Independence: observations are independent of each other
# For small n (like here, n=4 per group) these are hard to verify — interpret with caution.

import scipy.stats as stats

a = [40, 30, 35, 45]
b = [45, 35, 55, 25]
c = [55, 40, 30, 20]

f_stat, p_val = stats.f_oneway(a, b, c)
print(f'F-statistic: {f_stat:.4f}')
print(f'p-value:     {p_val:.4f}')

F-statistic: 0.1014
p-value:     0.9045


Since the calculated p-value (0.904)>0.05, we fail to reject the null hypothesis.There is no significant difference between the three treatments, at a 5% significance level.

## Two-way-ANOVA 

A botanist wants to know whether or not plant growth is influenced by sunlight exposure and watering frequency. She plants 30 seeds and lets them grow for two months under different conditions for sunlight exposure and watering frequency. After two months, she records the height of each plant, in inches.

In [2]:
import numpy as np
import pandas as pd

#create data
df = pd.DataFrame({'water': np.repeat(['daily', 'weekly'], 15),
                   'sun': np.tile(np.repeat(['low', 'med', 'high'], 5), 2),
                   'height': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})

In [3]:
df[:10]

,water,sun,height
0,daily,low,6
1,daily,low,6
2,daily,low,6
3,daily,low,5
4,daily,low,6
5,daily,med,5
6,daily,med,5
7,daily,med,6
8,daily,med,4
9,daily,med,5


In [4]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

#perform two-way ANOVA
model = ols('height ~ C(water) + C(sun) + C(water):C(sun)', data=df).fit()
sm.stats.anova_lm(model, typ=2)

,sum_sq,df,F,PR(>F)
C(water),8.533333,1.0,16.0000,0.000527
C(sun),24.866667,2.0,23.3125,0.000002
C(water):C(sun),2.466667,2.0,2.3125,0.120667
Residual,12.800000,24.0,NaN,NaN


# Analysis of Variance (ANOVA) Results

We can see the following p-values for each of the factors in the table:

- **Water:** p-value = 0.000527  
- **Sun:** p-value = 0.0000002  
- **Water * Sun:** p-value = 0.120667  

Since the p-values for water and sun are both less than 0.05, this means that both factors have a statistically significant effect on plant height.

And since the p-value for the interaction effect (0.120667) is not less than 0.05, this tells us that there is no significant interaction effect between sunlight exposure and watering frequency.

---
## ANOVA Assumption Checks

Before trusting the F-test result, verify the three key assumptions:
1. **Normality** — Shapiro-Wilk test per group
2. **Homogeneity of variance** — Levene's test
3. **Independence** — guaranteed by experimental design (check your study design)

In [5]:
import numpy as np
from scipy import stats

a = [40, 30, 35, 45]
b = [45, 35, 55, 25]
c = [55, 40, 30, 20]

# 1. Normality (Shapiro-Wilk per group)
print('Shapiro-Wilk Normality Test')
print('─' * 45)
for label, grp in [('Fertilizer A', a), ('Fertilizer B', b), ('Fertilizer C', c)]:
    stat, p = stats.shapiro(grp)
    verdict = 'OK' if p >= 0.05 else 'NON-NORMAL'
    print(f'  {label}: W={stat:.4f}, p={p:.4f}  ({verdict})')
print('  Note: With n=4 per group, normality tests have very low power.')
print('  Interpret results with caution for small samples.')
print()

# 2. Levene's test for equal variances
lev_stat, lev_p = stats.levene(a, b, c)
print('Levene\'s Test for Homogeneity of Variance')
print('─' * 45)
print(f'  W={lev_stat:.4f}, p={lev_p:.4f}')
if lev_p >= 0.05:
    print('  p >= 0.05 → No evidence of unequal variances. ANOVA assumption holds.')
else:
    print('  p < 0.05 → Significant variance inequality. Consider Welch ANOVA.')


Shapiro-Wilk Normality Test
─────────────────────────────────────────────
  Fertilizer A: W=0.9929, p=0.9719  (OK)
  Fertilizer B: W=0.9929, p=0.9719  (OK)
  Fertilizer C: W=0.9890, p=0.9523  (OK)
  Note: With n=4 per group, normality tests have very low power.
  Interpret results with caution for small samples.

Levene's Test for Homogeneity of Variance
─────────────────────────────────────────────
  W=1.3404, p=0.3094
  p >= 0.05 → No evidence of unequal variances. ANOVA assumption holds.


---
## Effect Size: η² (Eta-Squared)

The F-statistic tells us *whether* group means differ; **η²** tells us *how much* of
the total variance is explained by the group factor.

$$\eta^2 = \frac{SS_{between}}{SS_{total}}$$

Interpretation: η² < 0.06 = small, 0.06–0.14 = medium, ≥ 0.14 = large

In [6]:
import numpy as np
from scipy import stats

a = np.array([40, 30, 35, 45], dtype=float)
b = np.array([45, 35, 55, 25], dtype=float)
c = np.array([55, 40, 30, 20], dtype=float)

all_data   = np.concatenate([a, b, c])
grand_mean = all_data.mean()

ss_between = sum(len(g) * (g.mean() - grand_mean)**2 for g in [a, b, c])
ss_total   = sum((x - grand_mean)**2 for x in all_data)
eta_sq     = ss_between / ss_total

f_stat, p_val = stats.f_oneway(a, b, c)

print(f'F-statistic : {f_stat:.4f}')
print(f'p-value     : {p_val:.4f}')
print()
print(f'SS_between  : {ss_between:.4f}')
print(f'SS_total    : {ss_total:.4f}')
print(f'eta-squared : {eta_sq:.4f}')
if eta_sq < 0.06:
    size = 'small'
elif eta_sq < 0.14:
    size = 'medium'
else:
    size = 'large'
print(f'Effect size : {size} (eta^2={eta_sq:.3f})')


F-statistic : 0.1014
p-value     : 0.9045

SS_between  : 29.1667
SS_total    : 1322.9167
eta-squared : 0.0220
Effect size : small (eta^2=0.022)


---
## Post-Hoc Test: Tukey HSD

ANOVA's F-test only tells us *at least one group differs* — not *which pairs* differ.
**Tukey's Honestly Significant Difference (HSD)** controls the family-wise error rate
and identifies which specific group pairs have significantly different means.

> Note: With only 4 observations per group and a non-significant ANOVA (p=0.904),
> we don't expect any pair to be significant — but the method is demonstrated for
> completeness.

In [7]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import numpy as np

a = np.array([40, 30, 35, 45], dtype=float)
b = np.array([45, 35, 55, 25], dtype=float)
c = np.array([55, 40, 30, 20], dtype=float)

# Reshape for Tukey: (all values, group labels)
values = np.concatenate([a, b, c])
groups = ['A']*len(a) + ['B']*len(b) + ['C']*len(c)

tukey = pairwise_tukeyhsd(endog=values, groups=groups, alpha=0.05)
print(tukey)
print()
print('Reject column: True = that pair\'s means differ significantly at alpha=0.05')


 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     A      B      2.5 0.9534 -21.1704 26.1704  False
     A      C    -1.25 0.9881 -24.9204 22.4204  False
     B      C    -3.75 0.8989 -27.4204 19.9204  False
-----------------------------------------------------

Reject column: True = that pair's means differ significantly at alpha=0.05


---
## Non-Parametric Alternative: Kruskal-Wallis Test

When normality or variance homogeneity assumptions cannot be met, the
**Kruskal-Wallis test** is the non-parametric equivalent of one-way ANOVA.
It tests whether the group **distributions** differ using rank-based comparisons.

- No normality or equal-variance assumption required
- If significant: follow up with Dunn's test for pairwise comparisons
- Less powerful than ANOVA when normality truly holds

In [8]:
from scipy import stats
import numpy as np

a = [40, 30, 35, 45]
b = [45, 35, 55, 25]
c = [55, 40, 30, 20]

h_stat, p_kw = stats.kruskal(a, b, c)

print('Kruskal-Wallis Test (non-parametric ANOVA)')
print(f'H-statistic : {h_stat:.4f}')
print(f'p-value     : {p_kw:.4f}')
print()
if p_kw < 0.05:
    print('Reject H0: at least one group distribution differs.')
    print('Follow up with Dunn\'s test for pairwise comparisons.')
else:
    print('Fail to reject H0: no significant difference between group distributions.')
print()
print('Comparison with parametric ANOVA:')
f_stat, p_anova = stats.f_oneway(a, b, c)
print(f'  ANOVA   : F={f_stat:.4f}, p={p_anova:.4f}')
print(f'  Kruskal : H={h_stat:.4f}, p={p_kw:.4f}')
print('Both agree on the conclusion here (no significant difference).')


Kruskal-Wallis Test (non-parametric ANOVA)
H-statistic : 0.2447
p-value     : 0.8849

Fail to reject H0: no significant difference between group distributions.

Comparison with parametric ANOVA:
  ANOVA   : F=0.1014, p=0.9045
  Kruskal : H=0.2447, p=0.8849
Both agree on the conclusion here (no significant difference).


---
## ML/AI Connection

- **Comparing multiple models:** one-way ANOVA across $k$ algorithms' CV scores tests whether any algorithm differs significantly; Tukey HSD then identifies which pairs differ.
- **Feature importance via ANOVA F-statistic:** `sklearn.feature_selection.SelectKBest(f_classif)` ranks features by their F-statistic — the ratio of between-class to within-class variance.
- **Factorial experimental design** (two-way ANOVA) underlies ablation studies in ML: test the effect of two hyperparameters and their interaction simultaneously.
